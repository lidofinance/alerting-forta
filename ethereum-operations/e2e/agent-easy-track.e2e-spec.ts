import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-easy-track e2e tests", () => {
  let runBlock: (blockHashOrNumber: string | number) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(
        provideAgentPath("subagents/easy-track/agent-easy-track")
      ),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process tx with new motion created",
    async () => {
      const findings = await runTransaction(
        "0x6cead5592d65a47dcc099490db2e38b742860a47b04eb83718ca59a7bb1eb28c"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with granted role",
    async () => {
      const findings = await runTransaction(
        "0x55c4c7e33eb92da16871944879d52180c1a2e59c2701404abef864c5196ab7f1"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with revoked role",
    async () => {
      const findings = await runTransaction(
        "0x043e04411dd746562fd9c4244ac30570f09a737d73e9523edbdce722de3a2093"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with executed motion",
    async () => {
      const findings = await runTransaction(
        "0x9bbabb4891f324def5a0b073d89babaf7db84eb89d43afa46319ae7b377048c1"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with objected motion",
    async () => {
      const findings = await runTransaction(
        "0x3026189f7287678ca31403f77939ae812c9706eaca73d7966b405acebd56b2c4"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with granted role on RewardProgramsRegistry",
    async () => {
      const findings = await runTransaction(
        "0xd02f92e5a47792082f41ad6ca9e2d05d60f409a63328ac41a5d2704a7eb3fc1c"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );
});
