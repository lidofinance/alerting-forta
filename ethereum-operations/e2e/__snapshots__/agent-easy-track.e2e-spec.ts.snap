// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-easy-track e2e tests should process tx with executed motion 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "EASY-TRACK-MOTION-ENACTED",
  "description": "EasyTrack motion 362 was enacted",
  "labels": Array [],
  "metadata": Object {
    "args": "362",
  },
  "name": "‚úÖ EasyTrack: Motion executed",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-easy-track e2e tests should process tx with granted role 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "EASY-TRACK-ROLE-GRANTED",
  "description": "Role 0x0000000000000000000000000000000000000000000000000000000000000000 was granted to 0x2a61d3ba5030Ef471C74f612962c7367ECa3a62d on EasyTrack contract by 0x2a61d3ba5030Ef471C74f612962c7367ECa3a62d",
  "labels": Array [],
  "metadata": Object {
    "args": "0x0000000000000000000000000000000000000000000000000000000000000000,0x2a61d3ba5030Ef471C74f612962c7367ECa3a62d,0x2a61d3ba5030Ef471C74f612962c7367ECa3a62d",
  },
  "name": "üö® EasyTrack: Role was granted on EasyTrack contract",
  "protocol": "ethereum",
  "severity": 5,
  "type": 4,
}
`;

exports[`agent-easy-track e2e tests should process tx with granted role on RewardProgramsRegistry 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "REWARD-PROGRAMS-REGISTRY-ROLE-GRANTED",
  "description": "Role 0x0000000000000000000000000000000000000000000000000000000000000000 was granted by 0x2e59A20f205bB85a89C53f1936454680651E618e on RewardProgramsRegistry by 0x2a61d3ba5030Ef471C74f612962c7367ECa3a62d",
  "labels": Array [],
  "metadata": Object {
    "args": "0x0000000000000000000000000000000000000000000000000000000000000000,0x2e59A20f205bB85a89C53f1936454680651E618e,0x2a61d3ba5030Ef471C74f612962c7367ECa3a62d",
  },
  "name": "üö® Reward Programs: Role was granted on RewardProgramsRegistry",
  "protocol": "ethereum",
  "severity": 5,
  "type": 4,
}
`;

exports[`agent-easy-track e2e tests should process tx with new motion created 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "EASY-TRACK-MOTION-CREATED",
  "description": "New  motion [365](https://easytrack.lido.fi/motions/365) created by 0x834560F580764Bc2e0B16925F8bF229bb00cB759",
  "labels": Array [],
  "metadata": Object {
    "args": "365,0x834560F580764Bc2e0B16925F8bF229bb00cB759,0xBd2b6dC189EefD51B273F5cb2d99BA1ce565fb8C,0x000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000001000000000000000000000000834560f580764bc2e0b16925f8bf229bb00cb75900000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000096d0707a0aadf3200000,0x00000001231ac69a1a37649c6b06a71ab32ddd92158c80b800000024666712290000000000000000000000000000000000000000000096d0707a0aadf3200000b9e5cbb9ca5b0d659238807e84d0176930753d86000000c4f63648460000000000000000000000005a98fcbea516cf06857215779fd812ca3bef1b32000000000000000000000000834560f580764bc2e0b16925f8bf229bb00cb7590000000000000000000000000000000000000000000096d0707a0aadf32000000000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000001c4561737920547261636b3a20746f7020757020726563697069656e7400000000",
  },
  "name": "‚Ñπ EasyTrack: New motion created",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-easy-track e2e tests should process tx with objected motion 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "EASY-TRACK-MOTION-OBJECTED",
  "description": "EasyTrack motion 351 was objected by 0x22c05930160c6a74a6bA59436e8F7006176ed104",
  "labels": Array [],
  "metadata": Object {
    "args": "351,0x22c05930160c6a74a6bA59436e8F7006176ed104,10000000000000000000000,10000000000000000000000,0",
  },
  "name": "‚ÑπÔ∏è EasyTrack: Motion objected",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-easy-track e2e tests should process tx with revoked role 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "EASY-TRACK-ROLE-REVOKED",
  "description": "Role 0x9f959e00d95122f5cbd677010436cf273ef535b86b056afc172852144b9491d7 was revoked from 0x2a61d3ba5030Ef471C74f612962c7367ECa3a62d on EasyTrack contract by 0x2a61d3ba5030Ef471C74f612962c7367ECa3a62d",
  "labels": Array [],
  "metadata": Object {
    "args": "0x9f959e00d95122f5cbd677010436cf273ef535b86b056afc172852144b9491d7,0x2a61d3ba5030Ef471C74f612962c7367ECa3a62d,0x2a61d3ba5030Ef471C74f612962c7367ECa3a62d",
  },
  "name": "üö® EasyTrack: Role was revoked on EasyTrack contract",
  "protocol": "ethereum",
  "severity": 4,
  "type": 4,
}
`;
